src/Voron/Platform/Posix/Posix32BitsMemoryMapPager.cs

```diff
@@ -1,55 +1,577 @@
 using System;
+using System.Collections.Concurrent;
 using System.Collections.Generic;
-using System.ComponentModel;
 using System.Diagnostics;
 using System.IO;
-using System.IO.MemoryMappedFiles;
-using System.Runtime.CompilerServices;
 using System.Runtime.InteropServices;
-using Microsoft.Win32.SafeHandles;
+using System.Threading;
 using Sparrow;
+using Sparrow.Collections;
 using Sparrow.Logging;
 using Sparrow.Platform.Posix;
 using Sparrow.Utils;
 using Voron.Data;
 using Voron.Global;
 using Voron.Impl;
 using Voron.Impl.Paging;
-using Voron.Platform.Win32;
 
 namespace Voron.Platform.Posix
 {
     public unsafe class Posix32BitsMemoryMapPager : PosixAbstractPager
     {
+        private int _fd;
+        private readonly bool _isSyncDirAllowed;
+        private readonly bool _copyOnWriteMode;
+
+        private readonly ConcurrentDictionary<long, ConcurrentSet<MappedAddresses>> _globalMapping = new ConcurrentDictionary<long, ConcurrentSet<MappedAddresses>>(NumericEqualityComparer.Instance);
+        private long _totalMapped;
+        private int _concurrentTransactions;
+        private readonly ReaderWriterLockSlim _globalMemory = new ReaderWriterLockSlim();
+
+        private long _totalAllocationSize;
+        public const int AllocationGranularity = 64 * Constants.Size.Kilobyte;
+        private const int NumberOfPagesInAllocationGranularity = AllocationGranularity / Constants.Storage.PageSize;
+        public override long TotalAllocationSize => _totalAllocationSize;
+
         public Posix32BitsMemoryMapPager(StorageEnvironmentOptions options, string file, long? initialFileSize = null,
-                    bool usePageProtection = false) : base(options, usePageProtection)
+            bool usePageProtection = false) : base(options, usePageProtection)
+        {
+            FileName = file;
+
+            if (Options.CopyOnWriteMode)
+                ThrowNotSupportedOption(file);
+
+            _copyOnWriteMode = options.CopyOnWriteMode && file.EndsWith(Constants.DatabaseFilename);
+            _isSyncDirAllowed = PosixHelper.CheckSyncDirectoryAllowed(FileName);
+
+            PosixHelper.EnsurePathExists(FileName);
+
+            _fd = Syscall.open(file, OpenFlags.O_RDWR | OpenFlags.O_CREAT,
```

Comment:
@ayende Is it ok to set PosixSafeFlags on PosixMemMapPager ? we do not use it to write to the journal (as we do with PosixJournalWriter), only for reading from journal on recovery.  Also it will be wrong to use these flags on scratch/datafile pagers.. so.. ?